type ObjectValues<T extends object> = T[keyof T]

export const main = () => {
  /**
   * TypeScript Enums
   */
  const enums = () => {
    enum Response {
      No, // compiles to --> 1
      Yes // compiles to --> 2
    }

    const test = () => {
      // $ExpectError ðŸ’¥ NOPE  ðŸ¤¯???!!!
      const test: Response.No = 4

      function respond(recipient: string, message: Response) {
        /* ...implementation... */
      }

      // $ExpectError ðŸ’¥ NOPE  ðŸ¤¯???!!!
      respond('unknown', 4)
    }
  }

  /**
   * Idiomatic JS Enums (number) via object
   */
  const idiomaticEnums = () => {
    const Response = {
      No: 1,
      Yes: 2
    } as const

    // merge implementation with "Enum" typed literal
    type Response = ObjectValues<typeof Response>

    const test = () => {
      // $ExpectError ðŸ‘‰ âœ…
      const test: Response = 4

      function respond(recipient: string, message: Response) {
        /* ...implementation... */
      }

      // $ExpectError ðŸ‘‰ âœ…
      respond('unknown', 4)
    }
  }

  /**
   * TypeScript String Enums
   */
  const stringEnums = () => {
    enum Colors {
      Red = 'RED',
      Green = 'GREEN',
      Blue = 'BLUE'
    }

    const test = () => {
      // ERROR WAT ðŸ¤¯???
      const test: Colors.Red = 'RED'

      function favoriteColor(name: string, color: Colors) {
        /* ...implementation... */
      }

      // ERROR WAT ðŸ¤¯???
      favoriteColor('unknown', 'RED')
    }
  }

  /**
   * Idiomatic JS Enums (string) via object
   */
  const idiomaticStringEnums = () => {
    const Colors = {
      Red: 'RED',
      Green: 'GREEN',
      Blue: 'BLUE'
    } as const
    // merge implementation with "Enum" typed literal
    type Colors = ObjectValues<typeof Colors>

    const test = () => {
      // no ERROR âœ…
      const test: Colors = 'RED'

      function favoriteColor(name: string, color: Colors) {
        /* ...implementation... */
      }

      // no ERROR âœ…
      favoriteColor('unknown', 'RED')
    }
  }
}

// ======================
export const mainComplete = () => {
  /**
   * TypeScript Enums
   */
  const enums = () => {
    enum Response {
      No, // compiles to --> 1
      Yes // compiles to --> 2
    }

    const test = () => {
      // $ExpectError ðŸ’¥ NOPE  ðŸ¤¯???!!!
      const test: Response.No = 4
      // $ExpectError âœ…
      const test2: Response.Yes = Response.No
      // $ExpectError âœ…
      const test3: Response.Yes = 'hello'

      function respond(recipient: string, message: Response) {}

      // $ExpectError ðŸ’¥ NOPE  ðŸ¤¯???!!!
      respond('unknown', 4)
      respond('unknown', Response.No)
      respond('unknown', Response.Yes)
      // $ExpectError âœ…
      respond('unknown', 'ups')
    }
  }

  /**
   * Idiomatic JS Enums (number) via object
   */
  const idiomaticEnums = () => {
    const Response = {
      No: 1,
      Yes: 2
    } as const

    // merge implementation with "Enum" typed literal
    type Response = ObjectValues<typeof Response>

    const test = () => {
      // $ExpectError ðŸ‘‰ âœ…
      const test: Response = 4
      // NO ERROR âœ…
      const test2: Response = Response.No
      // $ExpectError âœ…
      const test3: Response = 'hello'

      function respond(recipient: string, message: Response) {}

      // $ExpectError ðŸ‘‰ âœ…
      respond('unknown', 4)
      respond('unknown', Response.Yes)
      respond('unknown', Response.No)
      respond('unknown', 1)
      // $ExpectError
      respond('unknown', 'ups')
    }
  }

  /**
   * TypeScript String Enums
   */
  const stringEnums = () => {
    enum Colors {
      Red = 'RED',
      Green = 'GREEN',
      Blue = 'BLUE'
    }

    const test = () => {
      // $ExpectError ðŸ¤¯???
      const test: Colors.Red = 'RED'
      // $ExpectError ðŸ¤¯???
      const test2: Colors = 'RED'
      const test3: Colors = Colors.Red

      function favoriteColor(name: string, color: Colors) {}

      // $ExpectError ðŸ¤¯???
      favoriteColor('unknown', 'RED')
      favoriteColor('unknown', Colors.Red)
      favoriteColor('unknown', Colors.Green)
    }
  }

  /**
   * Idiomatic JS Enums (string) via object
   */
  const idiomaticStringEnums = () => {
    const Colors = {
      Red: 'RED',
      Green: 'GREEN',
      Blue: 'BLUE'
    } as const
    // merge implementation with "Enum" typed literal
    type Colors = ObjectValues<typeof Colors>

    const test = () => {
      // ðŸ‘‰ yup no ERROR âœ…
      const test: Colors = 'RED'
      // ðŸ‘‰ yup no ERROR âœ…
      const test2: Colors = 'RED'
      const test3: Colors = Colors.Red

      function favoriteColor(name: string, color: Colors) {}

      // ðŸ‘‰ yup no ERROR âœ…
      favoriteColor('unknown', 'RED')
      favoriteColor('unknown', Colors.Red)
      favoriteColor('unknown', Colors.Green)
    }
  }
}

// ======================

const bad = () => {
  enum Response {
    No,
    Yes
  }

  enum Colors {
    Red = 'RED',
    Green = 'GREEN',
    Blue = 'BLUE'
  }

  function respond(recipient: string, message: Response) {}

  function favoriteColor(name: string, color: Colors) {}

  // TEST

  {
    // TypeScript Enums

    // $ExpectError ðŸ’¥ NOPE  !!!
    const test: Response.No = 4
    // $ExpectError
    const test2: Response.Yes = Response.No
    // $ExpectError
    const test3: Response.Yes = 'hello'

    // $ExpectError ðŸ’¥ NOPE  !!!
    respond('unknown', 4)
    respond('unknown', Response.No)
    respond('unknown', Response.Yes)
    // $ExpectError
    respond('unknown', 'ups')
  }

  {
    // TypeScript String Enums

    // $ExpectError
    const test: Colors.Red = 'RED'
    // $ExpectError
    const test2: Colors = 'RED'
    const test3: Colors = Colors.Red

    // $ExpectError
    favoriteColor('unknown', 'RED')
    favoriteColor('unknown', Colors.Red)
    favoriteColor('unknown', Colors.Green)
  }
}

const good = () => {
  type ObjectValues<T extends object> = T[keyof T]

  // $ExpectType { No: 1; Yes: 2; }
  const Response = {
    // we need to explicit cast values to get proper literal type
    No: 1,
    Yes: 2
  } as const
  // merge implementation with "Enum" typed literal
  // $ExpectType 1 | 2
  type Response = ObjectValues<typeof Response>

  // $ExpectType { Red: "RED"; Green: "GREEN"; Blue: "BLUE"; }
  const Colors = {
    Red: 'RED',
    Green: 'GREEN',
    Blue: 'BLUE'
  } as const
  // merge implementation with "Enum" typed literal
  // $ExpectType "RED" | "GREEN" | "BLUE"
  type Colors = ObjectValues<typeof Colors>

  function favoriteColor(name: string, color: Colors) {}

  function respond(recipient: string, message: Response) {}

  // TESTS

  {
    // Idiomatic JS Enums via object

    // $ExpectError ðŸ‘‰ yup catched ðŸ‘Œ
    const test: Response = 4
    // NO ERROR
    const test2: Response = Response.No
    // $ExpectError
    const test3: Response = 'hello'

    // $ExpectError ðŸ‘‰ yup catched ðŸ‘Œ
    respond('unknown', 4)
    respond('unknown', Response.Yes)
    respond('unknown', Response.No)
    respond('unknown', 1)
    // $ExpectError
    respond('unknown', 'ups')
  }

  {
    // Idiomatic JS Enums via object

    // ðŸ‘‰ yup no ERROR ðŸ‘Œ
    const test: Colors = 'RED'
    // ðŸ‘‰ yup no ERROR ðŸ‘Œ
    const test2: Colors = 'RED'
    const test3: Colors = Colors.Red

    // ðŸ‘‰ yup no ERROR ðŸ‘Œ
    favoriteColor('unknown', 'RED')
    favoriteColor('unknown', Colors.Red)
    favoriteColor('unknown', Colors.Green)
  }
}
